src/core/core.controller.js

```diff
@@ -446,8 +447,11 @@ helpers.extend(Chart.prototype, /** @lends Chart */ {
 				meta.controller = new ControllerClass(me, i);
 				newControllers.push(meta.controller);
 			}
+			sorted.push(meta);
```

Comment:
I wonder if we should check if the dataset is visible here so that we don't have to do it in `_getSortedDatasetMetas`

Reply:
Done

![image](https://user-images.githubusercontent.com/27971921/69014785-9f440180-0996-11ea-8156-6d63cfe4b282.png)

Reply:
I don't think we necessarily need a new property `visible` cause that's basically just `!hidden`, right? 

I was wondering if we could do the filtering of only visible datasets here to cache that as well, but on second look it seems like that's not possible because the legend shows the labels for visible and non-visible datasets

Reply:
https://github.com/chartjs/Chart.js/blob/eb0751c6d0c361a26ac05e431274aef9096b793c/src/core/core.controller.js#L897-L899

Reply:
I'm not sure about the use case where datasets are shared between charts, @etimberg any insights?

Reply:
yeah, that seems strange to me. perhaps that's a feature we can remove in 3.0?

Reply:
I think that was mainly for safety in case a user did it (it's somewhat annoying, especially for react, that we modify the data object).

Reply:
I'm not quite sure where we modify the data object, but stopping that sounds like another potentially good thing to change

Reply:
Perhaps 'modify' is the wrong word, but we attach extra info to the data object and so systems that treat that as immutable are unhappy https://github.com/chartjs/Chart.js/blob/e5b03a0975d57d7b32a9c31e62d145b8e79a2c7a/src/core/core.controller.js#L863

I used chart.js in a React project recently and tried writing my own wrapper and had a look at the source code of react-chartjs-2. They have to do a lot of work to handle the fact that the data object gets mutated. Dropping that would significantly simplify things

Reply:
I wonder if we need to treat `options` as immutable as well

Reply:
So the `meta.hidden` and `dataset.hidden` are both needed. `dataset.hidden` is provided in dataset configuration (loaded from storage maybe) and `meta.hidden` is used to override that, when clicking legend (avoiding touching the provided configuration).

Now this could be changed to read the dataset.hidden to meta.hidden at initializion and only use meta.hidden runtime. Provide hide() and show() for developer access. But not in the scope of this PR.

Reply:
Should `visible` be made private (i.e. `_visible`) since we may change it in the future? (maybe even worth adding a TODO to make the change you suggested)